### DISCUSSION TITLE
Albums Sorted AlbumArtist / Year

### DISCUSSION THREAD

**Discussion by clriis** [rern/rAudio#894] [Feature Requests]
An option to have albums sorted by year instead of alphabetically would be nice.
1'st sort key: AlbumArtist
2'nd sort key: Year

I'm aware that it can be obtained in file view by sorting the folders (adding year in front of album name), but I would like to have it the "Tag" views as well.

**Comment by gmarin37** [comment]
hi I use albumbyartist vue and the album are just sorted by artits but by date.
So I have done some modifications to have the right vue
1: /srv/http/bash/cmd-list.sh
##### normal list #############################################
listAll() {
        mpc -f '%album%^^[%albumartist%|%artist%]^^%file%' listall 2> /dev/null \
 go to 
##### normal list #############################################
listAll() {
        mpc -f '%album%^^[%albumartist%|%artist%]****_^^%date%**_**^^%file%' listall 2> /dev/null \

2: same file
               if [[ $albums ]]; then
                        for album in "${albums[@]}"; do
                                album_artist_file+=$( mpc -f '%album%^^[%albumartist%|%artist%]^^%file%' find album "$album" \

go to 
                        for album in "${albums[@]}"; do
                                album_artist_file+=$( mpc -f '%album%^^[%albumartist%|%artist%]_^^%date%_^^%file%' find 
album "$album" \

3: file /srv/http/bash/cmd-listsort.php
foreach( $lines as $line ) {
                        if ( substr( $line, -4 ) === '.cue' ) $line = dirname( $line );
                        $sort[] = stripSort( $line ).'^x^'.$line;
                        $data1 = explode( '^^', $line );
                        $line1 = $data1[ 1 ].'^^'.$data1[ 0 ].'^^'.$data1[ 2 ];
                        $sort1[] = stripSort( $line1 ).'^x^'.$line1;
                }
go to
foreach( $lines as $line ) {
                        if ( substr( $line, -4 ) === '.cue' ) $line = dirname( $line );
                        $sort[] = stripSort( $line ).'^x^'.$line;
                        $data1 = explode( '^^', $line );
                        $line1 = $data1[ 1 ]._'^^'.$data1[ 2 ]._'^^'.$data1[ 0 ].**_**'^^'.$data1[ 3 ]**_**;
                        $sort1[] = stripSort( $line1 ).'^x^'.$line1;
                }
4: file /srv/http/mpdlibrary.php
                foreach( $lists as $list ) {
                        $data      = explode( '^^', $list );
                        $index     = strtoupper( $data[ 0 ] );
                        $indexes[] = $index;
                        $path      = $data[ 3 ];

go to

                foreach( $lists as $list ) {
                        $data      = explode( '^^', $list );
                        $index     = strtoupper( $data[ 0 ] );
                        $indexes[] = $index;
                        $path      = $data[ 4 ];

5: same file
'<div class="coverart" data-index="'.$index.'">
        <a class="lipath">'.$path.'</a>
        <div><img class="lazyload" data-src="'.$coverfile.'^^^"></div>
        <span class="coverart1">'.$data[ 1 ].'</span>
        <gr class="coverart2">'.$space.'</gr>
</div>';

go to
'<div class="coverart" data-index="'.$index.'">
        <a class="lipath">'.$date_album.$path.'</a>
        <div><img class="lazyload" data-src="'.$coverfile.'^^^"></div>
        <span class="coverart1">'.$data[ 1 ].'</span>
        <gr class="coverart2">'.$space.'</gr>
        <gr class="coverart2">'.$data[ 3 ].'</gr>
</div>';

so now, after a refresh of your library, you have your cover, under one line with the artist, one line with the year and the last line is the name of the album

Don't forget to tag artist album and year tag on your files.
hope it's ok for you
regards
Gilbert

**Reply by gmarin37** [reply]
[cmd-list.sh.txt](https://github.com/rern/rAudio/files/11288193/cmd-list.sh.txt)
[cmd-listsort.php.txt](https://github.com/rern/rAudio/files/11288194/cmd-listsort.php.txt)
[mpdlibrary.php.txt](https://github.com/rern/rAudio/files/11288195/mpdlibrary.php.txt)

Here are the files
Regards
Gilbert

**Comment by clriis** [comment]
Hi,
Thanks for your efforts.
Unfortunately it does not work for me. After replacing the three files, the update process just continues indefinitely (36 hours+).
I even tried with a small size "test" library of 50 albums on a fresh install but to no avail.
I am using NAS/CIFS for music storage.

Regards,
Claus

**Reply by gmarin37** [reply]
hi I use Nas/cifs as you
have you check album by artist on the configration of library?
have you rename cmd-list.sh.txt to cmd-lisy.sh ? 
same for the other files, remove .txt
Gilbert

**Comment by gmarin37** [comment]
![coverart19](https://user-images.githubusercontent.com/3191979/233859742-2c315e06-197f-4f64-9a6b-a848452e8bb2.jpg)

**Comment by clriis** [comment]
Yes to all three points

**Comment by gmarin37** [comment]
![coverart20](https://user-images.githubusercontent.com/3191979/233860389-c4fac3ee-c14f-4747-917c-2e953d78b1c2.jpg)

**Comment by gmarin37** [comment]
[diff files.txt](https://github.com/rern/rAudio/files/11304683/diff.files.txt)

**Comment by gmarin37** [comment]
![coverart21](https://user-images.githubusercontent.com/3191979/233860832-8914677f-f04c-4b14-bcff-135a4bf382d1.jpg)
here is my system

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233860909-8809e019-e372-4a30-9483-3c769a7727c5.png)

**Comment by gmarin37** [comment]
can you give me one line of /srv/http/data/mpd/album and one line from /srv/http/data/mpd/albumbyartist?

**Comment by clriis** [comment]
2^^[2-fer]^^Mose Allison^^NAS/piCloud/Mose Allison/[2007] [2-fer]

1^^10,000 Maniacs

This is from an original install.

**Comment by gmarin37** [comment]
have tag all your files with the date?
Can you explain your directory structure?
NAS/piCloud/ is the share I suppose
fist level : artists
second level?

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233861709-1e096336-1816-44ae-a690-c7ee46c8f1e2.png)

SMB share on svr: piCloud is "music"

**Comment by gmarin37** [comment]
ok it's quite the same for me
can you test my files with only 10 albums on you library?
and about the date on the tags?

**Comment by clriis** [comment]
All files are tagged like this:

![image](https://user-images.githubusercontent.com/16999072/233862203-07695a24-a086-49c0-81c0-a2a9016603e6.png)

**Reply by gmarin37** [reply]
it seems ok

**Comment by gmarin37** [comment]
![coverart22](https://user-images.githubusercontent.com/3191979/233862304-09191b4a-71bf-4f1b-b0b2-5af329b2d069.jpg)
in red is important

**Comment by clriis** [comment]
Yes.
ALL albums have "Year" tagged.
I know because I used the tag to create the library structure

**Comment by clriis** [comment]
I can try with less albums. But how do you stop the update process and flush the MPD DB?
I got a process running now and it takes long because there are 30.000 tracks

**Reply by gmarin37** [reply]
can you ssh your player?
type ps aux | grep mpd and give me the answer

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233863029-bb80596d-2510-4b50-a6ba-7cc47111b982.png)

**Reply by gmarin37** [reply]
ls -la /srv/http/data/mpd

**Comment by clriis** [comment]
SORRY

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233863100-8b5b53eb-8c47-499c-a65c-02d2a7f03ba0.png)

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233863300-b96f665b-b471-4dfc-98d0-1d2742cee17e.png)

**Reply by gmarin37** [reply]
rm -rf /srv/http/data/mpd/updating
and hard shutdown your player
change the size of your library
restart your player
change with my files
make an full update

**Comment by clriis** [comment]
There are now only 6 albums on the target share.

I've replaced the original files and restarted the update process as you instructed, but it is still running.

![image](https://user-images.githubusercontent.com/16999072/233865094-b37f60e3-7e50-4fda-bd9e-32c598a5e373.png)

![image](https://user-images.githubusercontent.com/16999072/233865260-3e9eedd2-d8f2-4897-b1c4-374dac1e711b.png)

**Reply by gmarin37** [reply]
I don't understand why
how long does it takes with the normal files?

**Reply by gmarin37** [reply]
is there the /srv/http/data/mpd/updating file?

**Comment by clriis** [comment]
It takes about 3-4 hours with 30.000 tracks.

The updating file was there, but even though I removed it the updating process now seems to continue when looking at the web interface.

![image](https://user-images.githubusercontent.com/16999072/233866062-ed15feba-c28b-494b-bd27-986943a899ab.png)

It stopped when I was using the original files

**Reply by gmarin37** [reply]
what is the result of:
mpc -f '%album%^^[%albumartist%|%artist%]^^%file%^^%date%' listall

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233866462-8c1342d1-a9f3-4f6e-8d61-8e661919df58.png)

**Reply by gmarin37** [reply]
on the file file cmd-list.sh, can you give me the fonction listall (normally lines 36 to 43)

**Comment by clriis** [comment]
listAll() {
	mpc -f '%album%^^[%albumartist%|%artist%]^^%date%^^%file%' listall 2> /dev/null \
							| awk -F'/[^/]*$' 'NF && !/^\^/ {print $1}' \
							| sort -u
#	-F'/[^/]*$' - truncate %file% to path without filename
#	NF          - not empty lines
#	!/^\^/      - not lines with no album name
}

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233866753-cf73d7d3-02e7-446b-8174-0173c3c7d571.png)

**Reply by gmarin37** [reply]
lines 72 to 78

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233866999-7bf72960-02ae-40f2-a7fc-fa9ff317e061.png)

**Reply by gmarin37** [reply]
ok 
lines 124 to 130 on cmd-listsort.php

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233867258-cac4598f-4867-44d9-ad20-3b60804f9ac2.png)

**Reply by gmarin37** [reply]
ok is the update finish?
is there the file updating?

**Reply by clriis** [reply]
NO on both counts

**Comment by clriis** [comment]
We are talking about your files?

**Reply by gmarin37** [reply]
my files if they are on your player

**Reply by clriis** [reply]
Yes. Just wanted to make sure

**Reply by gmarin37** [reply]
ok is the update finish?
is there the file updating?

**Reply by clriis** [reply]
Not finished and no updating file (I did delete it earlier as I explained)

**Reply by gmarin37** [reply]
can you list the files in /srv/http/data/mpd ?
At what time did you stop update?

**Comment by gmarin37** [comment]
where do you live?

**Reply by clriis** [reply]
In Denmark, near the German border

**Reply by gmarin37** [reply]
ok i'm from France 
can we stop for now and talk again tomorrow?

**Comment by clriis** [comment]
![image](https://user-images.githubusercontent.com/16999072/233867489-6e33897e-648e-435b-973c-2af5825209b9.png)

I removed the file around 40 minutes ago.

**Reply by gmarin37** [reply]
lot of files are missing

![coverart23](https://user-images.githubusercontent.com/3191979/233867619-7935f53a-abd6-402a-9ac9-fb8cd43de3d5.jpg)

**Reply by clriis** [reply]
But that is presumably because it is a fresh install?

**Comment by clriis** [comment]
Yes.
And thank you so very much.
Claus

**Comment by clriis** [comment]
Here are my original files:

[mpdlibrary.php.orig.txt](https://github.com/rern/rAudio/files/11305043/mpdlibrary.php.orig.txt)
[cmd-list.sh.orig.txt](https://github.com/rern/rAudio/files/11305044/cmd-list.sh.orig.txt)
[cmd-listsort.php.orig.txt](https://github.com/rern/rAudio/files/11305045/cmd-listsort.php.orig.txt)

**Comment by gmarin37** [comment]
ok put the original files and make an update with light library
how is the directory /srv/http/data/mpd

**Comment by clriis** [comment]
Hi Gilbert,

I discovered last night that the updated image with the original files on a fresh install got exactly the same problem.

This morning I installed image 20221117 and it updated without problems in less than two hours for my full library.

Now I have replaced the files again and currently running an update. 

I will let you know when it's done.

Regards,
Claus

**Reply by gmarin37** [reply]
ok

**Comment by clriis** [comment]
It has now been updating since 12.40 and still not complete. Interestingly, I now get the Library page looking like this:

![image](https://user-images.githubusercontent.com/16999072/234036217-180fd4b8-d336-4eea-bd92-d85110d4f25a.png)

I don't have any DAB radio installed.

![image](https://user-images.githubusercontent.com/16999072/234036913-31f5dcb4-34e5-4ff6-9b62-1ba1f1ebc696.png)

**Reply by gmarin37** [reply]
i'm sorry but I don't have any more ideas to debug your player
we have different installations but I don't see why my code don't work on your player
the file cmd-list make the list of the files and fill the file album with all the albums : I just add the date on the lines 
the file cmd-listsort make a sort on the file album and fill albumbyartist with sorted lines artist-album : I just add the sort artist-year-album
the file mpdlivbrary just format the data to be seen: I just make 3 lines : artist+date+album

one more diff: I use flac files
I'm realy sorry to can't help you more
Gilbert

**Comment by clriis** [comment]
Hi Gilbert,
It is working now! Thank you so much for patience and effort.
I believe the reason for it not working with the 20221117 image must be compatibility issues.
I reinstalled the newest image and kept an eye on the updating process, remounting the share several times.

![image](https://user-images.githubusercontent.com/16999072/234376764-fc7121f2-b64d-43be-88ed-acb2c0e3d193.png)

**Reply by gmarin37** [reply]
great news
i'm very happy for you
Best regards 
Gilbert

**Reply by gmarin37** [reply]
Hi you need to redo the modifications on the next update
Gilbert

**Comment by clriis** [comment]
Hi Gilbert,

The new update seems to ad something to your files.

Have you got an update?

**Reply by gmarin37** [reply]
Hi Claus
I have done the update and redone the modifications and all is ok
No problem for my install

**Reply by rern** [reply]
Try a new option in Library to display Artist > Year > Album sort.

**Reply by clriis** [reply]
Thank you, @rern
